linera-execution/src/pricing.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module contains types related to fees and pricing.
+
+use async_graphql::InputObject;
+use linera_base::data_types::{Amount, ArithmeticError};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// A list of prices for different cost categories that each block causes for the validators.
+#[derive(Eq, PartialEq, Hash, Clone, Debug, Default, Serialize, Deserialize, InputObject)]
+pub struct Pricing {
+    /// The base price for each certificate, to compensate for the communication and signing
+    /// overhead.
+    pub certificate: Amount,
+    /// The price per unit of fuel used when executing effects and operations for user applications.
+    pub fuel: Amount,
+    /// The cost to store a block's operations and incoming messages, per byte.
+    pub storage: Amount,
+    /// The cost to store and send cross-chain messages, per byte.
+    pub messages: Amount,
+}
+
+impl Pricing {
+    pub fn certificate_price(&self) -> Amount {
+        self.certificate
+    }
+
+    pub fn messages_price(&self, data: &impl Serialize) -> Result<Amount, PricingError> {
+        let size =
+            u128::try_from(bcs::serialized_size(data)?).map_err(|_| ArithmeticError::Overflow)?;
+        Ok(self.messages.saturating_mul(size))
+    }
+
+    pub fn storage_price(&self, data: &impl Serialize) -> Result<Amount, PricingError> {
+        let size =
+            u128::try_from(bcs::serialized_size(data)?).map_err(|_| ArithmeticError::Overflow)?;
+        Ok(self.storage.saturating_mul(size))
+    }
+
+    pub fn fuel_price(&self, fuel: u64) -> Amount {
+        self.fuel.saturating_mul(u128::from(fuel))
+    }
+
+    /// Returns how much fuel can be paid with the given balance.
+    pub fn remaining_fuel(&self, balance: Amount) -> u64 {
+        u64::try_from(balance.saturating_div(self.fuel)).unwrap_or(u64::MAX)
+    }
+
+    #[cfg(any(test, feature = "test"))]
+    /// Creates a pricing with no cost for anything except fuel.
+    ///
+    /// This can be used in tests that need whole numbers in their chain balance and don't expect
+    /// to execute any Wasm code.
+    pub fn only_fuel() -> Self {
+        Pricing {
+            certificate: Amount::ZERO,
+            fuel: "0.0000001".parse().unwrap(),
```

Comment:
To remain consistent we should use underscores here? `0.000_000_1`

Reply:
Personally, I have never seen underscores in floats outside of Rust.

Reply:
I was just trying to make it easier to count the zeros. But if we get rid of the strings that's not a problem anyway. (Instead, you'll have to know the decimal metric prefixes up to "atto". :sweat_smile: )

Reply:
I think it helps in readability. I saw this quite a bit in the Java world (Post Java 7 IIRC).